nuxeo-features/nuxeo-platform-audit/nuxeo-platform-audit-core/src/test/java/org/nuxeo/ecm/platform/audit/AbstractAuditStorageTest.java

```diff
@@ -0,0 +1,171 @@
+import org.nuxeo.ecm.platform.audit.service.AuditBackend;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+/**
+ * Audit test that are common across all storage backends
+ */
+public abstract class AbstractAuditStorageTest {
+
+    public static final String ID_FOR_AUDIT_STORAGE_TESTS = "ID_FOR_AUDIT_STORAGE_TESTS";
+
+    public static final int NUM_OF_EVENTS = 56;
+
+    @Inject
+    protected AuditBackend auditBackend;
+
+    protected static ObjectMapper mapper = new ObjectMapper();
+
+    /**
+     * Sets up data for the tests below. Doesn't run using @Before so as not to interfere with other audit backends
+     */
+    public void setUpTestData() throws Exception {
+
+        AbstractAuditBackend backend = (AbstractAuditBackend) this.auditBackend;
+        AuditQueryBuilder builder = new AuditQueryBuilder().predicates(
+                Predicates.eq(LOG_EVENT_ID, ID_FOR_AUDIT_STORAGE_TESTS));
+        List<LogEntry> logs = backend.queryLogs(builder);
+
+        if (logs.isEmpty()) {
+
+            List<String> jsonEntries = new ArrayList<>();
+
+            for (int i = 1; i <= NUM_OF_EVENTS; i++) {
+                ObjectNode logEntryJson = mapper.createObjectNode();
+                logEntryJson.put(LOG_ID, 1000L+Integer.valueOf(i).longValue());
+                logEntryJson.put(LOG_EVENT_ID, ID_FOR_AUDIT_STORAGE_TESTS);
+                logEntryJson.put(LOG_DOC_PATH, i % 2 == 0 ? "/is/even" : "/is/odd");
+                jsonEntries.add(mapper.writeValueAsString(logEntryJson));
+            }
+
+            backend.append(jsonEntries);
+            flush();
+        }
+    }
+
+    protected abstract void flush() throws Exception;
+
+    @Test
+    public void testSaveAndScroll() throws Exception {
+        setUpTestData();
+        AbstractAuditBackend backend = (AbstractAuditBackend) this.auditBackend;
+
+        AuditQueryBuilder builder = new AuditQueryBuilder().predicates(
+                Predicates.eq(LOG_EVENT_ID, ID_FOR_AUDIT_STORAGE_TESTS));
+        checkNumberOfLoggedEvents(backend, builder);
+
+        ScrollResult<String> scrollResult = backend.scroll(builder, 5, 10);
+        int total = 0;
+        List<Long> ids = new ArrayList<>();
+
+        while (scrollResult.hasResults()) {
+            assertTrue(scrollResult.getResults().size() <= 5);
+            List<String> jsonEntries = scrollResult.getResults();
+            List<LogEntry> entries = jsonEntries.stream().map(json -> {
+                try {
+                    return mapper.readValue(json, LogEntryImpl.class);
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }).collect(Collectors.toList());
+            for (LogEntry entry : entries) {
+                assertEquals(ID_FOR_AUDIT_STORAGE_TESTS, entry.getEventId());
+                ids.add(entry.getId());
```

Comment:
I know that probably not your code. If you want to improve it, you can remove `entries` assignation and convert the for loop to a `forEach` call directly after the `map(json -> ...)`.